File: C/ShortURL.c
/*
 * ShortURL (https://github.com/delight-im/ShortURL)
 * Copyright (c) delight.im (https://www.delight.im/)
 * Licensed under the MIT License (https://opensource.org/licenses/MIT)
 */

/*
 * ShortURL: Bijective conversion between natural numbers (IDs) and short strings
 *
 * url_encode() takes an ID and turns it into a short string and writes it into a buffer
 * url_decode() takes a short string and turns it into an ID
 *
 * Features:
 * + large alphabet (51 chars) and thus very short resulting strings
 * + proof against offensive words (removed 'a', 'e', 'i', 'o' and 'u')
 * + unambiguous (removed 'I', 'l', '1', 'O' and '0')
 *
 * Example output:
 * 123456789 <=> pgK8p
 */

#include <string.h>
#include <limits.h>
#include <stdio.h>

const char alphabet[] = "23456789bcdfghjkmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ-_";
const size_t base = sizeof(alphabet) / sizeof(alphabet[0]) - 1;

// buff is a pre-allocated buffer provided by the caller, e.g. char[100]
// buffsize is the pre-allocated buffer's size, e.g. 100
size_t url_encode(char *buff, size_t buffsize, int num) {
	size_t idx = 0;
	size_t i;

	while (num > 0 && idx < buffsize) {
		buff[idx++] = alphabet[num % base];
		num /= base;
	}

	// if the buffer is not large enough
	if (idx == buffsize) {
		// fail
		return -1;
	}

	// add a null terminator
	buff[idx] = '\0';

	// reverse the encoded string
	for (i = 0; i < idx / 2; ++i) {
		char t = buff[i];
		buff[i] = buff[idx - i - 1];
		buff[idx - i - 1] = t;
	}

	// return the length of the encoded string
	return idx;
}

int url_decode(const char *url) {
	int num = 0;
	size_t idx, i, j, len;

	len = strlen(url);

	for (i = 0; i < len; ++i) {
		// locate the char in the alphabet
		for (j = 0; j <= base && alphabet[j] != url[i]; ++j);

		if (j > base) return INT_MAX;

		num = num * base + j;
	}

	return num;
}

File: Rust/shorturl.rs
/// # Shorturl
/// Bijective conversion between usize and string
///
/// short_url::encode(usize) takes an usize and turns it into a string
/// short_url::decode(String) takes a string and turns it into an usize
///
/// ## Features:
/// * Large alphabet (51 chars) and thus very short resulting strings
/// * Proof against offensive words (Removed 'a', 'e', 'i', 'o' and 'u')
/// * Unambiguous (Removed 'I', 'l', '1', 'O' and '0')
///
/// ## Example
/// * 123456789 <=> pgK8p
/// * 420 <=> bg
///
/// ShortURL (https://github.com/delight-im/ShortURL)
/// Copyright (c) andra.xyz (http://andra.xyz/)
/// Licensed under the MIT License (https://opensource.org/licenses/MIT)
mod short_url {
    static ALPHABET: &str = "23456789bcdfghjkmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ";
    static BASE: usize = 51;
    /// # Argument
    /// * `id` - A usize number that will be transformed
    pub fn encode(mut id: usize) -> String {
        let mut string: String = format!("");
        while id > 0 {
            string.push_str(&ALPHABET[(id % BASE)..(id % BASE + 1)]);
            id = id / BASE;
        }
        string.chars().rev().collect()
    }
    pub fn decode(string: String) -> usize {
        let mut number: usize = 0;
        for c in string.chars() {
            number = number * BASE + ALPHABET.find(c).unwrap();
        }
        number
    }
}
